diff --git a/binutils/objcopy.c b/binutils/objcopy.c
index 9732b86..dec0feb 100644
--- a/binutils/objcopy.c
+++ b/binutils/objcopy.c
@@ -2287,6 +2287,18 @@ copy_file (const char *input_filename, const char *output_filename,
 
       status = 1;
     }
+
+  if (status == 0)
+    {
+      struct stat statbuf;
+
+      /* No need to check the return value of stat().  It has already
+	 been checked in get_file_size().  */
+      stat (input_filename, &statbuf);
+
+      /* Try to preserve the permission bits.  */
+      chmod (output_filename, statbuf.st_mode);
+    }
 }
 
 /* Add a name to the section renaming list.  */

