--- a/ez-ipupdate.c	2003-05-27 17:48:26.000000000 +0300
+++ b/ez-ipupdate.c	2003-05-27 19:28:38.000000000 +0300
@@ -629,7 +629,7 @@
 #endif
   fprintf(stdout, "  -e, --execute <command>\tshell command to execute after a successful\n\t\t\t\tupdate\n");
   fprintf(stdout, "  -f, --foreground\t\twhen running as a daemon run in the foreground\n");
-  fprintf(stdout, "  -F, --pidfile <file>\t\tuse <file> as a pid file\n");
+  fprintf(stdout, "  -F, --pid-file <file>\t\tuse <file> as a pid file\n");
   fprintf(stdout, "  -g, --request-uri <uri>\tURI to send updates to\n");
   fprintf(stdout, "  -h, --host <host>\t\tstring to send as host parameter\n");
   fprintf(stdout, "  -i, --interface <iface>\twhich interface to use\n");
