diff -up expect-5.43/Makefile.in.libdir expect-5.43/Makefile.in
--- expect-5.43/Makefile.in.libdir	2004-05-07 20:10:30.000000000 +0200
+++ expect-5.43/Makefile.in	2008-09-25 12:24:08.000000000 +0200
@@ -317,7 +317,7 @@ EXP_AND_TK_LIBS = $(LDFLAGS) @EXP_AND_TK
 CFLAGS_INT = $(MH_CFLAGS) $(CPPFLAGS) $(XCFLAGS)
 
 LIB_INSTALL_DIR = $(tcl_libdir)
-LIB_RUNTIME_DIR = $(tcl_libdir)
+LIB_RUNTIME_DIR = $(subst $(INSTALL_ROOT),,$(tcl_libdir))
 # I don't understand why Tcl splits these up, but it does.  LIB_RUNTIME_DIR
 # can appear as part of the LD_SEARCH_FLAGS inherited by configure.
 
