diff -up tcl8.5.7/../unix/Makefile.in.old tcl8.5.7/unix/Makefile.in
--- tcl8.5.7/../unix//Makefile.in.old	2009-04-15 21:30:05.000000000 +0200
+++ tcl8.5.7/../unix/Makefile.in	2009-07-16 08:27:32.758444459 +0200
@@ -569,7 +569,7 @@ objs: ${OBJS}
 
 
 tclsh: ${TCLSH_OBJS} ${TCL_LIB_FILE}
-	${CC} ${CFLAGS} ${LDFLAGS} ${TCLSH_OBJS} @TCL_BUILD_LIB_SPEC@ ${LIBS} @EXTRA_TCLSH_LIBS@ \
+	${CC} ${CFLAGS} ${LDFLAGS} -fPIC ${TCLSH_OBJS} @TCL_BUILD_LIB_SPEC@ ${LIBS} @EXTRA_TCLSH_LIBS@ \
 		${CC_SEARCH_FLAGS} -o tclsh
 
 # Resetting the LIB_RUNTIME_DIR below is required so that the generated
