--- a/pptp_compat.c	2008/07/24 05:53:05	1.2
+++ b/pptp_compat.c	2008/07/25 00:13:56	1.3
@@ -8,12 +8,14 @@
 #include <sys/types.h>
 #include <unistd.h>
 #include <stdlib.h>
+#if defined (__SVR4) && defined (__sun) /* Solaris */
+#include <stropts.h>
+#endif
 #include <strings.h>
 #include "pptp_compat.h"
 #include <stdio.h>
 #include "util.h"
 
-
 #if defined (__SVR4) && defined (__sun) /* Solaris */
 /*
  * daemon implementation from uClibc
