diff -Nur module-init-tools-3.7-vanilla//Makefile.in module-init-tools-3.7/Makefile.in
--- module-init-tools-3.7-vanilla//Makefile.in	2009-04-09 00:00:26.000000000 +0200
+++ module-init-tools-3.7/Makefile.in	2009-04-23 13:27:13.000000000 +0200
@@ -35,7 +35,7 @@
 host_triplet = @host@
 target_triplet = @target@
 sbin_PROGRAMS = insmod$(EXEEXT) modprobe$(EXEEXT) rmmod$(EXEEXT) \
-	depmod$(EXEEXT) modinfo$(EXEEXT) insmod.static$(EXEEXT)
+	depmod$(EXEEXT) modinfo$(EXEEXT)
 bin_PROGRAMS = lsmod$(EXEEXT)
 noinst_PROGRAMS = modindex$(EXEEXT)
 subdir = .
@@ -69,11 +69,6 @@
 insmod_OBJECTS = $(am_insmod_OBJECTS)
 insmod_LDADD = $(LDADD)
 insmod_DEPENDENCIES =
-am_insmod_static_OBJECTS = insmod.$(OBJEXT)
-insmod_static_OBJECTS = $(am_insmod_static_OBJECTS)
-insmod_static_DEPENDENCIES =
-insmod_static_LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
-	$(insmod_static_LDFLAGS) $(LDFLAGS) -o $@
 am_lsmod_OBJECTS = lsmod.$(OBJEXT) util.$(OBJEXT) logging.$(OBJEXT)
 lsmod_OBJECTS = $(am_lsmod_OBJECTS)
 lsmod_LDADD = $(LDADD)
@@ -108,15 +103,13 @@
 CCLD = $(CC)
 LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
 SOURCES = $(depmod_SOURCES) $(EXTRA_depmod_SOURCES) $(insmod_SOURCES) \
-	$(EXTRA_insmod_SOURCES) $(insmod_static_SOURCES) \
-	$(EXTRA_insmod_static_SOURCES) $(lsmod_SOURCES) \
+	$(EXTRA_insmod_SOURCES) $(lsmod_SOURCES) \
 	$(EXTRA_lsmod_SOURCES) $(modindex_SOURCES) $(modinfo_SOURCES) \
 	$(EXTRA_modinfo_SOURCES) $(modprobe_SOURCES) \
 	$(EXTRA_modprobe_SOURCES) $(rmmod_SOURCES) \
 	$(EXTRA_rmmod_SOURCES)
 DIST_SOURCES = $(depmod_SOURCES) $(EXTRA_depmod_SOURCES) \
 	$(insmod_SOURCES) $(EXTRA_insmod_SOURCES) \
-	$(insmod_static_SOURCES) $(EXTRA_insmod_static_SOURCES) \
 	$(lsmod_SOURCES) $(EXTRA_lsmod_SOURCES) $(modindex_SOURCES) \
 	$(modinfo_SOURCES) $(EXTRA_modinfo_SOURCES) \
 	$(modprobe_SOURCES) $(EXTRA_modprobe_SOURCES) $(rmmod_SOURCES) \
@@ -245,15 +238,10 @@
 depmod_SOURCES = depmod.c util.c logging.c index.c moduleops.c tables.c zlibsupport.c config_filter.c util.h depmod.h logging.h index.h moduleops.h tables.h list.h testing.h  zlibsupport.h config_filter.h
 modinfo_SOURCES = modinfo.c util.c logging.c zlibsupport.c util.h testing.h zlibsupport.h
 modindex_SOURCES = modindex.c util.c logging.c index.c util.h logging.h index.h
-insmod_static_SOURCES = insmod.c
-insmod_static_LDFLAGS = -static
-# We don't want the $(zlib_flags) here: that makes a dynamic executable
-insmod_static_LDADD = 
 EXTRA_insmod_SOURCES = 
 EXTRA_lsmod_SOURCES = 
 EXTRA_modprobe_SOURCES = 
 EXTRA_rmmod_SOURCES = 
-EXTRA_insmod_static_SOURCES = 
 EXTRA_depmod_SOURCES = moduleops_core.c
 EXTRA_modinfo_SOURCES = 
 MAN5 = modprobe.conf.5 modules.dep.5 depmod.conf.5
@@ -357,9 +345,6 @@
 insmod$(EXEEXT): $(insmod_OBJECTS) $(insmod_DEPENDENCIES) 
 	@rm -f insmod$(EXEEXT)
 	$(LINK) $(insmod_OBJECTS) $(insmod_LDADD) $(LIBS)
-insmod.static$(EXEEXT): $(insmod_static_OBJECTS) $(insmod_static_DEPENDENCIES) 
-	@rm -f insmod.static$(EXEEXT)
-	$(insmod_static_LINK) $(insmod_static_OBJECTS) $(insmod_static_LDADD) $(LIBS)
 lsmod$(EXEEXT): $(lsmod_OBJECTS) $(lsmod_DEPENDENCIES) 
 	@rm -f lsmod$(EXEEXT)
 	$(LINK) $(lsmod_OBJECTS) $(lsmod_LDADD) $(LIBS)
diff -Nur module-init-tools-3.7-vanilla//configure module-init-tools-3.7/configure
--- module-init-tools-3.7-vanilla//configure	2009-04-09 00:00:09.000000000 +0200
+++ module-init-tools-3.7/configure	2009-04-23 13:23:39.000000000 +0200
@@ -2571,8 +2571,6 @@
 
 
 
-# If zlib is required, libz must be linked static, modprobe is in
-# /sbin, libz is in /usr/lib and may not be available when it is run.
 # Check whether --enable-zlib was given.
 if test "${enable_zlib+set}" = set; then
   enableval=$enable_zlib; if test "$enableval" = "yes"; then
@@ -2580,7 +2578,7 @@
 #define CONFIG_USE_ZLIB 1
 _ACEOF
 
-  zlib_flags="-Wl,-Bstatic -lz -Wl,-Bdynamic"
+  zlib_flags="-lz"
 fi
 fi
 
diff -Nur module-init-tools-3.7-vanilla//configure.ac module-init-tools-3.7/configure.ac
--- module-init-tools-3.7-vanilla//configure.ac	2009-04-08 23:54:15.000000000 +0200
+++ module-init-tools-3.7/configure.ac	2009-04-23 13:23:39.000000000 +0200
@@ -14,7 +14,7 @@
 [  --enable-zlib               Handle gzipped modules],
 [if test "$enableval" = "yes"; then
   AC_DEFINE(CONFIG_USE_ZLIB)
-  zlib_flags="-Wl,-Bstatic -lz -Wl,-Bdynamic"
+  zlib_flags="-lz"
 fi])
 
 AC_PROG_CC
